feature: Add `killthreads` command (closes #1580)

alufers opened the pull request:
This command allows the user to quickly kill multiple  threads by specyfying their ids as arguments to this command. It also supports the `--all` flag, which will kill overy thread except the currently selected one. This is useful for use with the `checkpoint` command.

The killing is done by calling `pthread_exit(0)`.

<!-- Please make sure to read the testing and linting instructions at https://github.com/pwndbg/pwndbg/blob/dev/DEVELOPING.md before creating a PR -->

alufers committed:
feature: Add `killthreads` command (closes #1580)

This command allows the user to quickly kill multiple  threads by
specyfying their ids as arguments to this command. It also supports
the `--all` flag, which will kill overy thread except the currently
selected one. This is useful for use with the `checkpoint` command.

The killing is done by calling `pthread_exit(0)`.

alufers committed:
fix: try fixing building test binaries by enabling -lpthread

alufers committed:
fix: remove error message check when calling pthread_exit

Removed the message check, because the error messages difffer between
versions of GDB.

alufers commented:
Cmon, the tests pass locally :D

disconnect3d commented on pwndbg/commands/killthreads.py:
```diff
@@ -0,0 +1,52 @@
+        if len(thread_ids) == 1:
+            print(message.success("Killed 1 thread"))
+        else:
+            print(message.success(f"Killed {len(thread_ids)} threads"))
```
Can we print the IDs of threads we kill?

disconnect3d commented on pwndbg/commands/killthreads.py:
```diff
@@ -0,0 +1,52 @@
+from pwndbg.commands import CommandCategory
+from pwndbg.gdblib.scheduler import lock_scheduler
+
+parser = argparse.ArgumentParser(description="Kill multiple threads at once.")
```
Can we add a long description (and make sure newline works here, I think some other command like vmmap have description with newlines) and inform users on:
```
Kill all or given threads.

Switches to given threads and calls pthread_exit(0) on them.
This is performed with scheduler-locking to prevent other threads from operating at the same time.

Killing all other threads may be useful to use GDB checkpoints, e.g., to test given input & restart the execution to the point of interest (checkpoint).
```

disconnect3d commented on pwndbg/commands/killthreads.py:
```diff
@@ -0,0 +1,52 @@
+                for thread in gdb.selected_inferior().threads()
+                if thread.num != current_thread_id
+            ]
+        for thread_id in thread_ids:
```
Should we maybe make sure that the provided thread ids are valid?

disconnect3d commented on pwndbg/commands/killthreads.py:
```diff
@@ -0,0 +1,52 @@
+        for thread_id in thread_ids:
+            gdb.execute(f"thread {thread_id}")
+            try:
+                gdb.execute("call (void) pthread_exit(0)")
```
Would calling all gdb.execute here with `to_string=True` prevent some GDB output from appearing?

alufers committed:
fix: Improve UX of the killthreads command

Add an extended description of the command, some validation on the thread IDs
and suppress GDB output.

alufers committed:
fix: lint

alufers committed:
fix: put the multiline help text in the correct place

disconnect3d commented on tests/gdb-tests/tests/test_command_killthreads.py:
```diff
@@ -0,0 +1,60 @@
+    assert len([thread for thread in gdb.selected_inferior().threads() if thread.num == 3]) == 0
+    assert len(gdb.selected_inferior().threads()) == initial_thread_count - 1
+
+    gdb.execute("kill")
```
Is this needed?

disconnect3d commented:
```
Launching pytest with args: ['/pwndbg/tests/gdb-tests/tests/test_command_killthreads.py::test_command_killthreads_kills_specific_thread', '-vvv', '-s', '--showlocals', '--color=yes']
============================= test session starts ==============================
platform linux -- Python 3.6.9, pytest-7.0.1, pluggy-1.0.0 -- /usr/bin/python
cachedir: .pytest_cache
rootdir: /pwndbg/tests/gdb-tests
collecting ... collected 1 item

tests/test_command_killthreads.py::test_command_killthreads_kills_specific_thread 

Program stopped.
0x00007ffff7dd4090 in _start () from /lib64/ld-linux-x86-64.so.2
Breakpoint 1 at 0x5555554007f6: file multiple_threads.c, line 26.
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7ffff77c2700 (LWP 1040)]
[New Thread 0x7ffff67c1700 (LWP 1041)]

Thread 1 "multiple_thread" hit Breakpoint 1, break_here () at multiple_threads.c:26
26	}
[Switching to thread 3 (Thread 0x7ffff67c1700 (LWP 1041))]
#0  clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:78

Thread 3 "multiple_thread" received signal SIGSEGV, Segmentation fault.
unwind_stop (version=<optimized out>, actions=<optimized out>, exc_class=<optimized out>, exc_obj=<optimized out>, context=0x7ffff67c0dc0, stop_parameter=0x0) at unwind.c:60
[Switching to thread 1 (Thread 0x7ffff7fe8740 (LWP 1037))]
#0  break_here () at multiple_threads.c:26
26	}
Killed threads with IDs: 3
FAILED
```

I think the test fails because the process gets SIGSEGV, but why does it get it? :/ Can it be that we call pthread_exit in a time when we should not?

disconnect3d commented:
@alufers can you try to fix the tests?

alufers committed:
Merge branch 'dev' into feature/killthreads-command

alufers committed:
tests: fix test failing due to a race condition when running in parrallel to other tests

Replaced asserts with loops which wait for a cundition to be met, so that the tests doesn't fail due to scheduling issues.

disconnect3d commented on tests/gdb-tests/tests/test_command_killthreads.py:
```diff
@@ -1,21 +1,37 @@
     gdb.execute("break break_here")
     gdb.execute("run")
-    assert len(gdb.selected_inferior().threads()) != 1
+    wait_until(lambda: len(gdb.selected_inferior().threads()) != 1)
```
We should probably just change the multiple_threads example so that the break_here is triggered only after other thread[s] are spawned. In other words, the other threads should synchronize with the main thread so that break_here is only called after we are in the desired state 😅

alufers replied:
I've added the semaphores, but it still fails, but later. I think there is a delay between when the thread dies and gdb reports it as gone

alufers committed:
tests: add more fixes for race conditions in test_killthreads

alufers committed:
fix: lint

alufers commented:
It works!

disconnect3d commented on pwndbg/commands/killthreads.py:
```diff
@@ -0,0 +1,74 @@
+@pwndbg.commands.OnlyWhenRunning
+def killthreads(thread_ids: Optional[List] = None, all: bool = False) -> None:
+    if len(thread_ids) == 0 and not all:
+        print(message.error("No thread IDs specified and --all not set"))
```
```suggestion
        print(message.error("No thread IDs or --all flag specified"))
```

disconnect3d committed:
Update pwndbg/commands/killthreads.py

alufers committed:
tests: Wait for exactly three threads
